pkg/git/poll/watch.go

```diff
@@ -0,0 +1,116 @@
+package poll
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"k8s.io/client-go/util/retry"
+
+	v1 "github.com/rancher/gitjob/pkg/apis/gitjob.cattle.io/v1"
+	"github.com/rancher/gitjob/pkg/git"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const defaultSyncInterval = 15
+
+type GitFetcher interface {
+	LatestCommit(ctx context.Context, gitjob *v1.GitJob, client client.Client) (string, error)
+}
+
+// Watch fetches the latest commit of all git repositories referenced by gitJobs with the syncInterval provided.
+type Watch struct {
+	gitJob  v1.GitJob
+	client  client.Client
+	done    chan bool
+	mu      *sync.Mutex
+	fetcher GitFetcher
+	log     logr.Logger
+}
+
+func NewWatch(gitJob v1.GitJob, client client.Client) Watcher {
+	return &Watch{
+		gitJob:  gitJob,
+		client:  client,
+		mu:      new(sync.Mutex),
+		fetcher: &git.Fetch{},
+		log:     ctrl.Log.WithName("git-latest-commit-poll-watch"),
+	}
+}
+
+// StartFetchingEverySyncIntervalInBackground fetches the latest commit every syncInternal in a goroutine.
+func (w *Watch) StartFetchingEverySyncIntervalInBackground(ctx context.Context) {
+	go w.startFetchingEverySyncInterval(ctx, time.NewTicker(time.Duration(calculateSyncInterval(w.gitJob))*time.Second))
+}
+
+// Finish stops watching for changes in the git repo.
+func (w *Watch) Finish() {
+	w.done <- true
```

Comment:
We cannot close the channel here because `Finish` might be called more than once (e.g when restarting because the `syncInterval` has changed). That will lead to a panic. Now that I look again at this it might make more sense to call this method `Stop` instead of `Finish`

Reply:
>  It is ok not to close it. Channel will be garbage collected

Oh, I didn't know that! I found [this](https://stackoverflow.com/a/8593986), and it makes sense that closing the channel is only a control signal and not necessary to be garbage collected.
Thanks!